[2.0.x] Cleanup/refactoring of EXTENSIBLE_UI

marcio-ao opened the pull request:
- Added axis_t, extruder_t, heater_t and fan_t to eliminate ambiguity, improve type safety.
- Regularized getter/setter argument order and naming.
- Tabular layout of "ui_api.h" to make it easier to use it as a reference.
- setAxisPosition now will not stack moves in buffer, allowing it to be
  called repeatedly on each touch ui tap.

Unambigous enum labels
----------------------------------
Methods which formerly used integers as arguments now use type-safe enum labels. The uses of labels also allows the use of BED instead of magical values such as zero.

An intentional choice was made to use C style "enums" rather than "class enums" to allow the brevity of writing "E0" rather than "extruder_t::E0"

Type-safety and compile-time checking:
----------------------------------------------------
There are now three new enum types: axis_t, heater_t, extruder_t and fan_t.

- axis_t: X, Y, Z
- extruder_t: E0, E1, E2, E3, E4, E5
- heater_t: H0, H1, H2, H3, H4, H5, BED
- fan_t: FAN0, FAN1, FAN2, FAN3, FAN4, FAN5

"axis_t" refer to a movement axis, but excludes extruders because there are operations which make sense for an extruder but not for a movement axis (such as linear advance). Separating "axis_t" and "extruder_t" allows methods to be written to take only one type while eliminating the need for run-time error checking or switch statements. Operations which apply to "axis_t" as well "extruder_t", such as setting positions, are handled via method overloading.

The additional "heater_t" type overlaps with "extruder_t", but also includes "BED", which has no associated stepper motor. Heaters are designated H0...H5. For convenience, methods which act on heaters also take "extruder_t", therefore "E" labels can be used interchangeably with "H" labels when setting temperatures.

Reordered and standardized getter/setter:
-------------------------------------------------------

The arguments to setter and getter methods have been reordered such that the value the setter is setting is always the first argument. The remaining arguments are modifiers and have been made identical between getter and setters, and the method names have been standardized, differing only by the "set" and "get" prefix. This allows for short-cuts via macros. For example, the standard way of increasing a temperature by 10 degrees looks like this:

```
UI::setTargetTemp_celsius(UI::getTargetTemp_celsius(E0) + 10, E0);
```

However, a variadic macro called UI_INCREMENT_BY has been created which expresses this more concisely as:

```
UI_INCREMENT_BY(TargetTemp_celsius, 10, E0)
```

A way forward for Ultralcd?
-------------------------------------------------------

This standardization of getter/setters also may allow more complex macros to be created, allowing for a possible future migration of the Ultralcd code to an EXTENSIBLE_UI module (if this is desired). This could be done by replacing a macro like this:

```
MENU_MULTIPLIER_ITEM_EDIT_CALLBACK(int3, MSG_NOZZLE, &thermalManager.target_temperature[0], 0, HEATER_0_MAXTEMP - 15, watch_temp_callback_E0);
```

With a macro like this:

```
MENU_MULTIPLIER_ITEM_EDIT_CALLBACK(int3, MSG_NOZZLE, watch_temp_callback_E0, TargetTemp_celsius, E0);
```

In this case, the temperature limits are removed (because such limits can be enforced by the setter and getters) and the pointer to a memory location is replaced with the setter/getter name (TargetTemp_celsius) and associated modifiers (E0).

Compatibility with earlier EXTENSIBLE_UI:
--------------------------------------------------------

The reordering of the parameters and the introduction of types will break code that formerly relied on the EXTENSIBLE_UI, but at this early stage of API development, I don't think anyone will be affected! The changes are fairly obvious and the added type-safety will ensure that the compiler will flag all code that needs to be modified. Apologies if this PR messes anyone up!

marcio-ao committed:
Cleanup of EXTENSIBLE_UI

- Added axis_t, extruder_t, heater_t and fan_t to eliminate ambiguity, improve type safety.
- Regularized getter/setter argument order and naming.
- Tabular layout of "ui_api.h" to make it easier to use it as a reference.
- setAxisPosition now will not stack moves in buffer, allowing it to be
  called repeatedly on each touch ui tap.

ejtagle commented:
@marcio-ao :To be honest, i really like the way EXTENSIBLE_UI is going.
But i don´t like the usage of macros. They are type-unsafe and make refactoring of the code a pain in the ***
Try CLion as development environment (recommended!), and you will see. 
But I agree that using templates, there is no way to create function identifiers from part of them...

marcio-ao commented:
> But i don´t like the usage of macros. They are type-unsafe and make refactoring of the code a pain in the ***

@ejtagle: My implementation of UI_INCREMENT_BY started out as a thought experiment to see if I could make my getter/setters macro friendly, because I knew this would pave the way for UltraLCD to use the API. But after I made the macro, I enjoyed the brevity it provided, so I decided to keep it.

I could add "increment" functions to each setter/getter pair, but I think the macro does the job without adding too much bulk. If C++ provided better way to synthesize code like macros did, I would use it.

thinkyhead committed:
Apply coding standards

thinkyhead committed:
Apply coding standards

thinkyhead committed:
Update ui_api.h

thinkyhead commented:
Yes, for the LCD and other output aimed at users we have historically used the string "E1" when referring to the extruder with index 0. But within the code —outside of printed strings— I think it's important to be consistent with the rest of the code and use "E0" for the first extruder. The same goes for "H1" and other indexes.

If anything, for consistency it would be better to alter the user-facing strings rather than create inconsistencies in the internal naming and usage.

marcio-ao commented:
@thinkyhead: I struggled with the pros and cons of changing the convention. But my rationale was that the EXTENSIBLE_UI's API provides a "firewall" between Marlin internal code and UI code that may be developed in parallel. Given that division, I thought the API was a reasonable place to make the change of convention.

One advantage about using one-based indexing is that if someone gets confused about what convention is being used and types E0, the compiler will yell at them, as E0 does not exist. On the other hand, if the first extruder is E0 and someone who mistakenly thinks E1 is the first extruder will write incorrect code that compiles. So the added degree of safety seemed like a point in favor of one-based labels.

I don't have a strong opinion on it, however, and could change the labels to be zero-based if you really prefer that.

thinkyhead commented on Marlin/src/lcd/extensible_ui/ui_api.cpp:
```diff
@@ -445,7 +495,8 @@ namespace UI {
   uint32_t getProgress_seconds_elapsed() {
-    return IFPC(print_job_timer.duration() / 1000UL, 0);
+    const duration_t elapsed = print_job_timer.duration();
+    return elapsed.value;
```
Thanks for the correction. I missed that the `duration()` method in `stopwatch.h` divides the current `millis_t` value by 1000 before returning it.

Are we guaranteed to have `print_job_timer` defined, if `printcounter.h` isn't included? Let's add some Travis CI tests for `EXTENSIBLE_UI` with and without all the extras so we can confirm everything compiles ok.

marcio-ao replied:
Perhaps add the header anyway?

thinkyhead commented on Marlin/src/lcd/extensible_ui/ui_api.h:
```diff
@@ -49,161 +49,211 @@ typedef const __FlashStringHelper *progmem_str;
+  /* Delay and timing routines */
+  /* Should be used by the EXTENSIBLE_UI to safely pause or measure time */
+  /* safe_millis must be called at least every 1 sec to guarantee time */
+  /* yield should be called within lengthy loops */
```
C-style per-line comments aren't necessary for Marlin code. Use C++ single line or use a single block style comment when the length over 2-3 lines.

marcio-ao replied:
Yeah. Not sure why I did it that way. Normally I use single block style comment. I generally like to keep the asterisks running all the way down the left, however since I feel it helps make the comment visually stand out.

thinkyhead commented on Marlin/src/lcd/extensible_ui/ui_api.h:
```diff
@@ -49,161 +49,211 @@ typedef const __FlashStringHelper *progmem_str;
+  void        onLoadSettings                 ();
 };
+
+/* Helper macros to increment or decrement a value. For example:
```
For consistency and for the sake of code-folding, please format block comments like so…
```cpp
/**
 * First line of comment
 *
 * More stuff
 */
```

marcio-ao replied:
Okay. I'll try to remember that. It's similar to the style I usually go for, except for the two leading asterisks.

marcio-ao replied:
Well shucks. The two leading asterisks breaks the syntax coloring in my editor (scite).

EDIT: Or maybe comments are supposed to be gray... they are green when I have only one asterisk. No idea what the color coding is **supposed** to be.

thinkyhead replied:
Some editors will distinguish between ordinary comments and Doxygen/HeaderDoc style comments.

thinkyhead replied:
Some editors will distinguish between ordinary comments and Doxygen/HeaderDoc style comments.

thinkyhead commented:
> I don't have a strong opinion on it, however, and could change the labels to be zero-based if you really prefer that.

I've stated the case as it stands now. The only place we use "E1" to refer to the first extruder is in serial and LCD output. I don't know why this convention has been used since Marlin's inception, since most board silkscreens refer to the first extruder as "E0," but that's how it's been. Unless we're going to change E0 to E1 everywhere, I think it's vital to stick with zero-based indexing and not make this code an outlier. If consistency is very important, the better route would be to change the LCD display and serial output (assuming this won't confuse hosts) to use zero-based indexing instead. However, I assume users are now used to "E1" always referring to the first extruder.

marcio-ao commented:
> If consistency is very important, the better route would be to change the LCD display and serial output (assuming this won't confuse hosts) to use zero-based indexing instead.

Our usability folks have made me change it on our printers, when E0 showed up in the LCD. We techies can do it either way, but muggles don't count from zero :)

marcio-ao commented:
In an ideal world, I would prefer it if Marlin used zero valued integers for the first extruder, but had an enum that would label that as E1, so it would be the same label as the LCD. But the config files do have tons of places where E0 appears and gcode itself seems to be zero-based, so certainly there's a lot in favor of keeping it zero based. I'll change it on the API on Monday.

marcio-ao committed:
Implemented better manual moves for EXTENSIBLE_UI (#12205)

- Calling setAxisPosition_mm will no longer buffer the entire move to the new position, but will instead
  cause small moves towards it to be made during the idle loop. This allows the user to adjust the
  destination even after the move has started and makes the UI feel much more responsive.
- As suggested by @ejtagle, the new code keeps the planner buffer full to ensure smooth motion without stops and gos.

marcio-ao committed:
Changed En, Hn and FANn to zero-based indices.

- Labels are now consistent with the reset of Marlin code.

marcio-ao committed:
Updated comment style.

marcio-ao committed:
Fix FAN index.

thinkyhead committed:
Update ui_api.cpp

thinkyhead committed:
No need to align things

thinkyhead committed:
Update ui_api.h

thinkyhead committed:
Update ui_api.cpp

Roxy-3D commented:
> In an ideal world, I would prefer it if Marlin used zero valued integers for the first extruder

I agree...   I suggest we consider doing this at the next major release.   The GCode numbers need to match what is displayed on the LCD Panel. 

> I'll change it on the API on Monday.

All fine and good....   But let's actually merge it into the code base at a major release.    Let's not do this mid-stream.  It is too disruptive.

InsanityAutomation commented:
I'll put my vote here as well, seen confusion in t0 and t1 a few too many times troubleshooting on Facebook groups...

marcio-ao commented:
> I agree... I suggest we consider doing this at the next major release. The GCode numbers need to match what is displayed on the LCD Panel.

You are saying make the first extruder be labeled E0 on the LCD panel? Please don't do this. It would be a major headache for manufacturers such as ourselves who already have printed documentation and even labels on the printers themselves. If I have to change all these labels on every release of Marlin, I'll quit my job and go grow tomatoes.

@Roxy-3D: You are actually disagreeing with me. I would like to see E1 used for the 1st extruder throughout Marlin. This can be an enum with value zero, but E0 should never be presented to the user.

marcio-ao commented:
> I'll put my vote here as well, seen confusion in t0 and t1 a few too many times troubleshooting on Facebook groups...

Switching to zero-based indices won't clarify anything, as long as the majority of English language speakers count from 1.

InsanityAutomation commented:
I'd just like to see consistency either direction. As it sits, slicers and gcode are all 0 and 1, firmware code is 0 and 1, LCD is the only place it's 1 and 2.

As much as I like tomato based things from ketchup to pizza, I'll at least try to keep you from jumping off that bridge!

marcio-ao commented:
> As it sits, slicers and gcode are all 0 and 1, firmware code is 0 and 1, LCD is the only place it's 1 and 2.

Not true. I've been looking at screenshots of slicers and printers online:
- Ultimaker Cura: "1" and "2"   (https://i.ytimg.com/vi/a80v66d0pAw/maxresdefault.jpg)
- Slic3r: "1" and "2" (http://manual.slic3r.org/expert-mode/images/multipleextruders/printer_settings_general_multiple_extruder_options.png)
- Simplify3D: "Left extruder" and "Right extruder" (https://www.simplify3d.com/wp-content/uploads/2014/04/Dual-Extrusion-Wizard-300x217.png)
- Mattercontrol: 1 and 2 (https://lh3.ggpht.com/j8gwWbcTi0eW8VVSX1LmtRRcYsSnedLaNwDD6xzCXifBRtOu4ztRwanlLUstRS-Bw-avelNsRswPK1sin9dZMkiXzA)
- Octo-print: T0 and T1 (http://zeeproforums.technobly.com/uploads/default/original/1X/59c46a885123bc99489f64f4459e828b0c673003.png)
- Prusa MMU: Extruder 1, 2, 3, 4 (https://www.prusa3d.com/downloads/manual/prusa3d_manual_mk2mm_en.pdf?2, page 18)
- Due Panel: 1, 2, 3, 4, 5, 6 (https://miscsolutions.files.wordpress.com/2014/11/2015-11-25-20-30-13-2.jpg)
- Raise3D LCD: L and R (https://cdn.shopify.com/s/files/1/1069/4816/files/Print-Resuming.jpg?8782614013645265209)

Almost everyone, with the exception of Octoprint, has moved away from the zero-based numbering scheme.

marcio-ao commented:
It is extremely confusing to a user when their software has buttons for 1 and 2, and this does not match what their printer is telling them. Most users don't look at GCODE files directly, so GCODE can remain zero-based, but for everywhere a non-expert user is able to see, it should be one based.

ejtagle commented:
All enums and code should be 0-based. But messages on LCD should be 1-based. 
The 0-based thing is C inheritance. In C, arrays are 0 based, and that is why it makes sense to keep that
Human Users are 1-based, so it makes sense Messages should be 1-based ;)

marcio-ao commented:
> All enums and code should be 0-based. But messages on LCD should be 1-based.

This is the Marlin status quo.

marcio-ao commented:
@ejtagle : I'll give you one thumbs up... or should that be zero thumbs up?

ejtagle commented:
The other (strange) option would be to let users specify a different name for each extruder (configurable via config_adv.h), so, for example, if user wants to name extruders in a different way, it could be done.
For example "Left extruder"/"Right extruder".... or even more creative naming, as assigning them human names.... ;)

InsanityAutomation commented:
The section in s3d mentioned is a variable like that. Taz dual profile actually says front and rear. Everywhere except the profile name is a 0 based list though where you select tool number.

marcio-ao commented:
> For example "Left extruder"/"Right extruder".... or even more creative naming, as assigning them human names.... ;)

Certainly in the case of UltraLCD, which already has localization strings, it would be easy to make strings corresponding to each extruder. This would also solve the problem where "E" as an abbreviation for extruder doesn't make sense for a particular language.

So perhaps the language files should declare constants like this:

```
#define MSG_E0                   _UxGT("E0")
#define MSG_E1                   _UxGT("E1")
...
#define MSG_FAN0               _UxGT("FAN0")
#define MSG_FAN1               _UxGT("FAN1")
```

This would allow for extruders to be named "Alice" and "Bob" and the fans "Fan of Alice" and "Fan of Bob", if that's what someone wants :)

ejtagle commented:
Exactly @marcio-ao ... Or label extruders in the printer and use those labels to reference them in the LCD.
